security/src/main/java/io/airlift/security/cert/CertificateBuilder.java

```diff
@@ -0,0 +1,262 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.security.cert;
+
+import io.airlift.security.der.DerUtils;
+
+import javax.security.auth.x500.X500Principal;
+
+import java.io.ByteArrayInputStream;
+import java.net.InetAddress;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.Signature;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.security.der.DerUtils.encodeBooleanTrue;
+import static io.airlift.security.der.DerUtils.encodeContextSpecificSequence;
+import static io.airlift.security.der.DerUtils.encodeContextSpecificTag;
+import static io.airlift.security.der.DerUtils.encodeInteger;
+import static io.airlift.security.der.DerUtils.encodeNull;
+import static io.airlift.security.der.DerUtils.encodeOctetString;
+import static io.airlift.security.der.DerUtils.encodeOid;
+import static io.airlift.security.der.DerUtils.encodeSequence;
+import static io.airlift.security.der.DerUtils.encodeUtcTime;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.time.ZoneOffset.UTC;
+import static java.util.Objects.requireNonNull;
+
+public class CertificateBuilder
+{
+    private static final byte[] SHA_256_WITH_RSA_ENCRYPTION_OID = encodeOid("1.2.840.113549.1.1.11");
+    private static final byte[] SUBJECT_KEY_IDENTIFIER_OID = encodeOid("2.5.29.14");
+    private static final byte[] AUTHORITY_KEY_IDENTIFIER_OID = encodeOid("2.5.29.35");
+    private static final byte[] BASIC_CONSTRAINTS_OID = encodeOid("2.5.29.19");
+    private static final byte[] SUBJECT_ALT_NAME_OID = encodeOid("2.5.29.17");
+
+    private RSAPublicKey publicKey;
+    private RSAPrivateKey privateKey;
+
+    private long serialNumber;
+    private X500Principal issuer;
+    private Instant notBefore;
+    private Instant notAfter;
+    private X500Principal subject;
+    private final List<String> sanDnsNames = new ArrayList<>();
+    private final List<InetAddress> sanIpAddresses = new ArrayList<>();
+
+    private CertificateBuilder() {}
+
+    public static CertificateBuilder certificateBuilder()
+    {
+        return new CertificateBuilder();
+    }
+
+    public CertificateBuilder setKeyPair(KeyPair keyPair)
+    {
+        requireNonNull(keyPair, "keyPair is null");
+        checkArgument(keyPair.getPublic() instanceof RSAPublicKey && keyPair.getPrivate() instanceof RSAPrivateKey, "KeyPair must contain RSA keys");
```

Comment:
Maybe do
```java
checkArgument(keyPair.getPublic() instanceof RSAPublicKey, "not an RSA key: %s", keyPair.getPublic());
checkArgument(keyPair.getPrivate() instanceof RSAPrivateKey, "not an RSA key: %s", keyPair.getPrivate());
```
So you know the actual type (or just let the cast fail which should include the type)